File: JavaPools/PoolsLibrary/src/main/java/ru/urfu/javapools/poolslibrary/controller/PoolException.java
package ru.urfu.javapools.poolslibrary.controller;

@SuppressWarnings("serial")
public class PoolException extends Exception {

	private Object _key;
	
	public PoolException (Object key, String message) {
		this(key, message, null);
	}
	
	public PoolException (Object key, String message, Throwable cause) {
		super(message, cause);
		_key = key;
	}
	
	public Object getKey() {
		return _key;
	}
}

File: CSharpPools/PoolsLibrary/Controller/PoolException.cs
using System;

namespace PoolsLibrary.Controller
{
    public class PoolException<TK> : Exception
    {
        public TK Key { get; private set; }

        public PoolException(TK key, string message = null, Exception innerException = null)
            : base(message,innerException)
        {
            Key = key;
        }

        public PoolException(string message = null, Exception innerException = null)
            : this (default(TK), message, innerException)
        {
        }
    }
}
